custom_components/gtfs_rt/PublicTransportData.py

```diff
@@ -0,0 +1,382 @@
+                url=self._trip_update_url,
+                headers=self._headers,
+                label="trip data",
+            )
+            if entity.HasField("trip_update")
+        ]
+
+        trip_update_df = gtfs_data_to_df(feed_entities, label="trip")
+        debug_dataframe(trip_update_df, "Trip Update")
+
+        # create vehicle info dataframe
+        v_feed_entities = get_gtfs_feed_entities(
+            url=self._vehicle_position_url,
+            headers=self._headers,
+            label="vehicle positions",
+        )
+        vehicle_info_df = gtfs_data_to_df(v_feed_entities, label="vehicle")
+        debug_dataframe(vehicle_info_df, "Vehicle Info")
+
+        # add vehicle to trip update info TODO: fix merging- some missing?
+        trip_update_df = pd.merge(
+            left=trip_update_df, right=vehicle_info_df, how="left"
+        )
+        debug_dataframe(trip_update_df, "Merged Trip & Vehicle Info")
+
+        # get static timetable data
+        timetable_df = StaticMasterGTFSInfo(
+            url=self.static_gtfs_url, CachedData=self.CachedGTFSData
+        ).departure_info
+        debug_dataframe(timetable_df, "Static Timetable Info")
+
+        # merge live and static data
+        trip_update_df = pd.merge(
+            trip_update_df,
+            timetable_df,
+            how="outer",
+            on="trip_id",
+            indicator="source",
+        )
+
+        # convert categorical columns to strings
+        trip_update_df[
+            [
+                "stop_id_x",
+                "stop_id_y",
+                "route_id_x",
+                "route_id_y",
+                "direction_id_x",
+                "direction_id_y",
+            ]
+        ] = trip_update_df[
+            [
+                "stop_id_x",
+                "stop_id_y",
+                "route_id_x",
+                "route_id_y",
+                "direction_id_x",
+                "direction_id_y",
+            ]
+        ].astype(
+            str
+        )
+        # choose non-NA values for stop_id, route_id, and direction_id
+        # combine stop_id, route_id, and direction_id columns
+        trip_update_df["stop_id"] = trip_update_df["stop_id_x"].combine_first(
+            trip_update_df["stop_id_y"]
+        )
+        trip_update_df["route_id"] = trip_update_df[
+            "route_id_x"
+        ].combine_first(trip_update_df["route_id_y"])
+        trip_update_df["direction_id"] = trip_update_df[
+            "direction_id_x"
+        ].combine_first(trip_update_df["direction_id_y"])
+
+        # drop intermediate columns
+        trip_update_df = trip_update_df.drop(
+            columns=[
+                "stop_id_x",
+                "stop_id_y",
+                "route_id_x",
+                "route_id_y",
+                "direction_id_x",
+                "direction_id_y",
+            ]
+        )
+        debug_dataframe(trip_update_df, "Merged live and static info")
+
```

Comment:
Need to allow for delays here. If it's delayed, we want to keep it until the updated arrival time is in the past.

Reply:
Resolved by moving this until after the updated_arrival_time is generated.